### DISCUSSION TITLE
RDkit Conformer generator

### DISCUSSION THREAD

**Discussion by chem555** [rdkit/rdkit#6047] [Q&A]
I wish to ask whether the RDKit conformer generator is able to handle reliably organic chloride salts as input please?

**Comment by greglandrum** [comment]
@chem555 can you please provide an example, as either SMILES or a mol/SDF file, of the kind of molecule you mean?

**Reply by chem555** [reply]
Yes sure, organic salts as the following: [Cl-].NC1=NC2=C([NH+]=C[N]2COC(CO)CO)C(=O)N1

**Reply by ShangChien** [reply]
According to the definition of smiles, `[Cl-].NC1=NC2=C([NH+]=C[N]2COC(CO)CO)C(=O)N1` is consist of two molecules. `[Cl-]` and `NC1=NC2=C([NH+]=C[N]2COC(CO)CO)C(=O)N1` are split by symbol `.` .
RDKIT can't guess  the position of `Cl-`,  and  in rdkit no function can directly get conformer by pass param two molecule. But, you can do it like this:

- remove `Cl-`

```
#test
mol_test=Chem.MolFromSmiles('NC1=NC2=C([NH+]=C[N]2COC(CO)CO)C(=O)N1')
mol_test = Chem.AddHs(mol_test)
cids=rdDistGeom.EmbedMultipleConfs(mol_test,numConfs=50, params=etkdg)
##view the 3D model of conformers
viewer = py3Dmol.view(width=300, height=300)
viewer.addModel(Chem.MolToMolBlock(mol_test, confId=2),'mol') #confId = [0:49]
viewer.setStyle({"stick":{}})
viewer.zoomTo()
viewer.show()
```
![image](https://user-images.githubusercontent.com/48350636/216867806-f49eadf6-fa14-4131-91d1-6ff22c49ed5b.png)

-   OR : if you sure Cl is connect to N+ (caution to adjust charge to make sure `Chem.AddHs(mol_test)` work well )

```
#test
mol_test=Chem.MolFromSmiles('NC1=NC2=C([N+](Cl)=C[N]2COC(CO)CO)C(=O)N1')
mol_test = Chem.AddHs(mol_test)
cids=rdDistGeom.EmbedMultipleConfs(mol_test,numConfs=50, params=etkdg)
##view the 3D model of conformers
viewer = py3Dmol.view(width=300, height=300)
viewer.addModel(Chem.MolToMolBlock(mol_test, confId=2),'mol')
viewer.setStyle({"stick":{}})
viewer.zoomTo()
viewer.show()
```
![image](https://user-images.githubusercontent.com/48350636/216867488-3b73893c-4c6b-4075-84e5-618e9c6c2e05.png)

**Comment by greglandrum** [comment]
@chem555 If a molecule has multiple fragments (as your example does), the RDKit conformer generator will by default generate the coordinates for the fragments separately and then combine them. So atoms will generally be too close to each other. Here's a simple example of that:
```
In [9]: m = Chem.AddHs(Chem.MolFromSmiles('[Na+].[OH-]'))

In [10]: params = rdDistGeom.ETKDGv3()

In [11]: params.randomSeed = 0xf00d

In [12]: rdDistGeom.EmbedMolecule(m,params)
Out[12]: 0

In [13]: print(Chem.MolToMolBlock(m))

     RDKit          3D

  3  1  0  0  0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 Na  0  0  0  0  0 15  0  0  0  0  0  0
    0.4939    0.0000    0.0000 O   0  0  0  0  0  1  0  0  0  0  0  0
   -0.4939    0.0000    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  2  3  1  0
M  CHG  2   1   1   2  -1
M  END
```
Notice that the Na is actually *between* the O and the H.

You can tell the code to use all the atoms at once when generating the conformer, but since it doesn't know anything about charged interactions, it will just try to ensure that the fragments are not touching each other. In this case that result in the pieces being VERY far apart:
```
In [14]: params.embedFragmentsSeparately = False

In [15]: rdDistGeom.EmbedMolecule(m,params)
Out[15]: 0

In [16]: print(Chem.MolToMolBlock(m))

     RDKit          3D

  3  1  0  0  0  0  0  0  0  0999 V2000
  297.7843    0.6244    0.0000 Na  0  0  0  0  0 15  0  0  0  0  0  0
 -148.4021    0.2058    0.0000 O   0  0  0  0  0  1  0  0  0  0  0  0
 -149.3822    0.2196    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  2  3  1  0
M  CHG  2   1   1   2  -1
M  END
```

**Reply by greglandrum** [reply]
An advanced topic: if you know which charged atoms you want to pair and have close to each other, you can modify the bounds matrix which describes how far about atoms should be:
```
In [17]: bm = rdDistGeom.GetMoleculeBoundsMatrix(m)

In [18]: bm[0,1]
Out[18]: 1000.0

In [19]: bm[1,0]
Out[19]: 3.95

In [20]: bm[1,0] = 2.5

In [21]: bm[0,1] = 3.5

In [22]: from rdkit import DistanceGeometry

In [23]: DistanceGeometry.DoTriangleSmoothing(bm)
Out[23]: True

In [24]: params.SetBoundsMat(bm)

In [25]: rdDistGeom.EmbedMolecule(m,params)
Out[25]: 0

In [26]: print(Chem.MolToMolBlock(m))

     RDKit          3D

  3  1  0  0  0  0  0  0  0  0999 V2000
    2.3372    0.6244    0.0000 Na  0  0  0  0  0 15  0  0  0  0  0  0
   -0.8871   -0.1880    0.0000 O   0  0  0  0  0  1  0  0  0  0  0  0
   -1.4501    0.6134    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  2  3  1  0
M  CHG  2   1   1   2  -1
M  END
```
If you're interested in trying something like this and understanding what's going on, I'd encourage taking a quick look at the RDKit documentation about the conformer generator (https://www.rdkit.org/docs/GettingStartedInPython.html#working-with-3d-molecules) and reading some of the included references.

**Reply by chem555** [reply]
Thank you!
